specutils/analysis/width.py

```diff
@@ -0,0 +1,212 @@
+        Region within the spectrum to calculate the gaussian sigma width.
+
+    Returns
+    -------
+    full_width : `~astropy.units.Quantity` or list (based on region input)
+        Approximate full width of the signal
+    """
+    return _computation_wrapper(_compute_gaussian_sigma_width, spectrum, region)
+
+
+def gaussian_fwhm(spectrum, region=None):
+    """
+    Estimate the full width half max of the spectrum assuming it is gaussian.
+    This will be calculated over the regions, if they are specified.
+
+    Parameters
+    ----------
+    spectrum : `~specutils.spectra.spectrum1d.Spectrum1D`
+        The spectrum object overwhich the width will be calculated.
+
+    region: `~specutils.utils.SpectralRegion` or list of `~specutils.utils.SpectralRegion`
+        Region within the spectrum to calculate the FWHM value.
+
+    Returns
+    -------
+    gaussian_fwhm : `~astropy.units.Quantity` or list (based on region input)
+        Approximate full width of the signal at half max
+    """
+    return _computation_wrapper(_compute_gaussian_fwhm, spectrum, region)
+
+
+def fwhm(spectrum, region=None):
+    """
+    Compute the true full width half max of the spectrum.
+    
+    This makes no assumptions about the shape of the spectrum (e.g. whether it
+    is Gaussian). It finds the maximum of the spectrum, and then locates the
+    point closest to half max on either side of the maximum, and
+    measures the distance between them. This will be calculated over the
+    regions, if they are specified. 
+
+    Parameters
+    ----------
+    spectrum : `~specutils.spectra.spectrum1d.Spectrum1D`
+        The spectrum object over which the width will be calculated.
+
+    region: `~specutils.utils.SpectralRegion` or list of `~specutils.utils.SpectralRegion`
+        Region within the spectrum to calculate the FWHM value.
+
+    Returns
+    -------
+    whm : `~astropy.units.Quantity` or list (based on region input)
+        Full width of the signal at half max
+    """
+    return _computation_wrapper(_compute_fwhm, spectrum, region)
+
+
+def _compute_gaussian_fwhm(spectrum, region=None):
+    """
+    Estimate the full width of the spectrum at half max.
+
+    Parameters
+    ----------
+    spectrum : `~specutils.spectra.spectrum1d.Spectrum1D`
+        The spectrum object overwhich the width will be calculated.
+
+    region: `~specutils.utils.SpectralRegion`
+        Region within the spectrum to calculate the FWHM value.
+
+    Returns
+    -------
+    gaussian_fwhm : `~astropy.units.Quantity` or list (based on region input)
+        Approximate full width of the signal at half max
+
+    Notes
+    -----
+    This is a helper function for the above `gaussian_fwhm()` method.
+
+    """
+
+    if region is not None:
+        calc_spectrum = region.extract(spectrum)
+    else:
+        calc_spectrum = spectrum
+
+    flux = calc_spectrum.flux
+    spectrum = calc_spectrum.spectral_axis
+
+    dx = spectrum - np.mean(spectrum)
```

Comment:
Piggy-backing on the comment above in the docs: I think this is the magic part that could make it relatively insensitive to the centering.  If instead of doing ``np.mean`` here, you instead do ``np.mean(spectrum*flux)/np.sum(flux)``, you get get a centroid estimate of the center of the line.

Alternatively, you can just use the centroid function *directly* from `specutils.analysis.centroid` - that might not be quite as efficient (it has to do the region-extraction twice), but is a better re-use of code.

One step even further is to change this line to ``dx = spectrum - center_func(calc_spectrum)`` and have ``center_func`` be a user-choosable keyword, which you would have default to `specutils.analysis.centroid`.

(You've basically been getting my stream-of-conciousness here... But TL;DR I think you should do what I said in the last paragraph unless it's going to take too long)

Reply:
This makes sense. The duplication of region extraction occurred to me too, but this is probably something we should handle when addressing #266.